streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -121,21 +121,32 @@ private void runOnce() throws InterruptedException {
         }
 
         private void performActionsOnTasks() {
-            tasksAndActionsLock.lock();
-            try {
-                for (final TaskAndAction taskAndAction : getTasksAndActions()) {
+            while (!tasksAndActions.isEmpty()) {
+                Task toInitialize = null;
+                tasksAndActionsLock.lock();
+                try {
+                    final TaskAndAction taskAndAction = tasksAndActions.remove();
                     final Action action = taskAndAction.getAction();
                     switch (action) {
                         case ADD:
-                            addTask(taskAndAction.getTask());
+                            final Task task = taskAndAction.getTask();
+                            final Task existingTask = updatingTasks.putIfAbsent(task.id(), task);
+                            if (existingTask != null) {
+                                throw new IllegalStateException((existingTask.isActive() ? "Active" : "Standby") + " task " + task.id() + " already exist, " +
+                                    "should not try to add another " + (task.isActive() ? "active" : "standby") + " task with the same id. " + BUG_ERROR_MESSAGE);
+                            }
+                            toInitialize = task;
```

Comment:
Could you extract this code to a method that returns a boolean that can be assigned to `toInitialize`? I think that makes the code more readable.

Reply:
boolean won't quite work because then I don't have a reference to the task in the initialization block. 

I introduced a function for adding the task though.

Reply:
Yeah, sorry! My bad! I meant more or less what you did, actually!